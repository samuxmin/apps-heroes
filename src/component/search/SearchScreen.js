import React, { useMemo } from 'react'
import queryString from 'query-string'
import { useLocation } from 'react-router-dom';
import {heroes} from '../../data/heroes.js'
import { useForm } from '../../hooks/useForm.js';
import { HeroCard } from '../heroes/HeroCard.js';

export const SearchScreen = ({history}) => {


    const {q = ''} = queryString.parse(useLocation().search)

    const [ values, handleInputChange] = useForm({search:q})
    const { search } = values
    
    const getHeroesByName = ( name ) => {
        if(!name){return []}
        
        name = name.toLowerCase();
        return heroes.filter(hero=> hero.superhero.toLowerCase().includes(name))
    }
    const heroesFiltered = useMemo(()=>getHeroesByName(q),[q]);
    

    const handleSearch = (e) =>{
        e.preventDefault()
        history.push(`?q=${search}`)
    }
    return (
        <>
            <h1>Search</h1>
            <hr />
            <div className="search-container">
                <div className="search">
                    <hr />
                    <form onSubmit={handleSearch}>
                        <input 
                            name='search'
                            type='text'
                            placeholder='Find your hero'
                            className='search-input'
                            onChange={handleInputChange}
                            value={search}
                            autoComplete='off'
                            />
                            <button
                                type='submit'
                                className='btn btn-search'
                            >Search...</button>
                    </form>
                </div>
                <div className="search-results">
                    <h2>Results</h2>
                    <hr />
                    {(q==='') 
                    && 
                    <div className="aviso"
                        >Search a hero
                    </div>}
                    {
                        (q!=="" && heroesFiltered.length=== 0) && <div className="aviso">No hero found with {q}</div>
                    }
                    {
                        heroesFiltered.map(hero =>(<HeroCard key={hero.id} {...hero}/>))
                    }
                </div>
            </div>
        </>
    )
}
