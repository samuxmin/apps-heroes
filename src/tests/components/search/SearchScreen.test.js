import React from "react";
import { mount } from "enzyme";
import { MemoryRouter, Route } from "react-router-dom";
import { SearchScreen } from "../../../component/search/SearchScreen";
describe("Pruebas en <SearchScreen />", () => {
  test("debe de mostrarse correctamente con valores por defecto", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search"]}>
        <Route path="/search" component={SearchScreen} />
      </MemoryRouter>
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(".aviso").text().trim()).toBe("Search a hero");
  });
  test("debe de mostrar al heroe y el input con el valor del queryString", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search?q=batman"]}>
        <Route path="/search" component={SearchScreen} />
      </MemoryRouter>
    );
    expect(wrapper.find("input").prop("value")).toBe("batman");
    expect(wrapper.find("HeroCard").exists()).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });

  test("debe de mostrar un error si no se encuentra el hero", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search?q=toretto"]}>
        <Route path="/search" component={SearchScreen} />
      </MemoryRouter>
    );
    expect(wrapper.find(".aviso").text().trim()).toBe(
      "No hero found with toretto"
    );
    expect(wrapper.find("HeroCard").exists()).toBe(false);
  });

  test("debe de llamar el push del history", () => {
    const history = { push: jest.fn() };
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search"]}>
        <Route
          path="/search"
          component={() => <SearchScreen history={history} />}
        />
      </MemoryRouter>
    );
    wrapper.find("input").simulate("change", {
      target: {
        name: "search",
        value: "batman",
      },
    });
    wrapper.find("form").prop("onSubmit")({
      preventDefault() {},
    });
    expect(history.push).toHaveBeenCalledWith('?q=batman')
  });
});
