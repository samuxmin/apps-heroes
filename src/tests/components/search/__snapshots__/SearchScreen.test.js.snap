// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchScreen /> debe de mostrar al heroe y el input con el valor del queryString 1`] = `
Array [
  <h1>
    Search
  </h1>,
  <hr />,
  <div
    className="search-container"
  >
    <div
      className="search"
    >
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="search-input"
          name="search"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batman"
        />
        <button
          className="btn btn-search"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="search-results"
    >
      <h2>
        Results
      </h2>
      <hr />
      <div
        className="card-container"
      >
        <div
          className="card"
        >
          <div
            className="img-container"
          >
            <img
              alt="Batman"
              src="./assets/heroes/dc-batman.jpg"
            />
          </div>
          <div
            className="card-text"
          >
            <h3>
              Batman
            </h3>
            <br />
            <p>
              Bruce Wayne
            </p>
            <br />
            <small>
              Detective Comics #27
            </small>
            <br />
            <a
              className="btn"
              href="/hero/dc-batman"
              onClick={[Function]}
            >
              Mas...
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`Pruebas en <SearchScreen /> debe de mostrarse correctamente con valores por defecto 1`] = `
Array [
  <h1>
    Search
  </h1>,
  <hr />,
  <div
    className="search-container"
  >
    <div
      className="search"
    >
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="search-input"
          name="search"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          className="btn btn-search"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="search-results"
    >
      <h2>
        Results
      </h2>
      <hr />
      <div
        className="aviso"
      >
        Search a hero
      </div>
    </div>
  </div>,
]
`;
